internal/controller/sync_controller.go

```diff
@@ -0,0 +1,237 @@
+	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	// "github.com/vmware-tanzu/velero/pkg/persistence"
+	// "github.com/vmware-tanzu/velero/pkg/plugin/clientmgmt"
+	// "github.com/vmware-tanzu/velero/pkg/plugin/clientmgmt/process"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	// "k8s.io/apimachinery/pkg/util/sets"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	nacv1alpha1 "github.com/migtools/oadp-non-admin/api/v1alpha1"
+	"github.com/migtools/oadp-non-admin/internal/common/constant"
+	"github.com/migtools/oadp-non-admin/internal/common/function"
+)
+
+// NonAdminBackupSyncReconciler reconciles a NonAdminBackupStorageLocation object
+type NonAdminBackupSyncReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	OADPNamespace string
+	SyncPeriod    time.Duration
+}
+
+type nacCredentialsFileStore struct{}
+
+func (f *nacCredentialsFileStore) Path(*corev1.SecretKeySelector) (string, error) {
+	return "/tmp/credentials/secret-file", nil
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state,
+// defined in NonAdminBackupStorageLocation object Spec.
+func (r *NonAdminBackupSyncReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	// on NonAdminBackupStorageLocation create/update and every configured timestamp
+	logger.V(1).Info("NonAdminBackup Synchronization start")
+
+	nonAdminBackupStorageLocation := &nacv1alpha1.NonAdminBackupStorageLocation{}
+	err := r.Get(ctx, req.NamespacedName, nonAdminBackupStorageLocation)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(1).Info(err.Error())
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "Unable to fetch NonAdminBackupStorageLocation")
+		return ctrl.Result{}, err
+	}
+
+	// Get related Velero BSL
+	if nonAdminBackupStorageLocation.Status.VeleroBackupStorageLocation == nil {
+		err = fmt.Errorf("VeleroBackupStorageLocation not yet processed")
+		logger.Error(err, "VeleroBackupStorageLocation not yet processed")
+		return ctrl.Result{}, err
+	}
+	backupStorageLocation, err := function.GetVeleroBackupStorageLocationByLabel(
+		ctx, r.Client, r.OADPNamespace,
+		nonAdminBackupStorageLocation.Status.VeleroBackupStorageLocation.NACUUID,
+	)
+	if err != nil {
+		logger.Error(err, "Failed to get VeleroBackupStorageLocation")
+		return ctrl.Result{}, err
+	}
+	if backupStorageLocation == nil {
+		err = fmt.Errorf("VeleroBackupStorageLocation not yet created")
+		logger.Error(err, "VeleroBackupStorageLocation not yet created")
+		return ctrl.Result{}, err
+	}
+
+	// // OPTION 1: copy how Velero does
+	// // PROBLEM: NAC Pod will not have Velero Pod files
```

Comment:
We would need to mount secret and plugins to NAC Pod to allow this approach, right?

I do not know how much work would that be

Reply:
I think the sync controller should restore the NAB objects based on the Backup objects that are found in the OADP namespace (by default `openshift-adp`). I am not sure if the sync controller should really go into s3 BSLs itself?

I do not know why there is anything with secrets to be done, or I am missing something.

Once the BSL is in the OADP namespace the secret is also in that namespace and corresponds to the BSL and not the NABSL.

Reply:
NaSYNC : openshift-adp <-> non-admin ns
Velero Sync: s3 <-> openshift-adp